### PR TITLE
hwmv2: intel_adsp: Port to HWMv2

### PR DESCRIPTION
**Pull request by golowanow**
Port `intel_adsp` SoC family: `ace` and `cavs` SoC series, and related board configurations:

* `intel_adsp_cavs25`
* `intel_adsp_cavs25_tgph` changed to `intel_adsp_cavs25/intel_tgl_adsp/tgph`
* `intel_adsp_ace15_mtpm`
* `intel_adsp_ace20_lnl`

cc:
- #51831

### CODE REVIEW

**Comment by golowanow** [comment]
@tejlmand, @nashif, @dcpleung  - for your early feedback as a draft.

**Comment by golowanow** [comment]
1. adjusted to #68346 
2. `intel_adsp_cavs25_tgph` board configuration is converted to a HWMv2 revision `intel_adsp_cavs25/intel_tgl_adsp/tgph`

File: soc/intel/intel_adsp/cavs/_soc_inthandlers.h
Comment (golowanow) [inline soc/intel/intel_adsp/cavs/_soc_inthandlers.h:1]: this move from `intel_adsp/common/include` to `intel_adsp/cavs/` resolves build issue when this header file was applied with the HWMv2 for `adsp` builds as well

File: boards/intel/intel_adsp_cavs25/board.yml
Comment (golowanow) [inline boards/intel/intel_adsp_cavs25/board.yml:1]: do we really need to change `tgph` to `intel_adsp_cavs25` board variant ?

**Comment by nashif** [comment]
> * `intel_adsp_cavs25_tgph` changed to `intel_adsp_cavs25/intel_tgl_adsp/tgph`

oh wow, not nice at all, @tejlmand, we need away for this to be shorter and to the point, lots of redundancy in that name.

**Comment by nordicjm** [comment]
> > * `intel_adsp_cavs25_tgph` changed to `intel_adsp_cavs25/intel_tgl_adsp/tgph`
> 
> oh wow, not nice at all, @tejlmand, we need away for this to be shorter and to the point, lots of redundancy in that name.

As said before, if there is only one soc then you can omit the soc, i.e. `intel_adsp_cavs25//tgph`, that should only be omitted by users on the command line though, .yaml files and sample.yml/testcase.yml should use the full name

File: soc/intel/intel_adsp/Kconfig [hunk 1]
```diff
@@ -1,22 +1,17 @@
...
 
 if SOC_FAMILY_INTEL_ADSP
 
-config SOC_FAMILY
-	string
-	default "intel_adsp"
-
 # Select SoC Part No. and configuration options
-source "soc/soc_legacy/xtensa/intel_adsp/*/Kconfig.soc"
+# rsource "*/Kconfig.soc"
```
Comment (nordicjm) [inline soc/intel/intel_adsp/Kconfig:14]:
```suggestion
rsource "*/Kconfig"
```

File: soc/intel/intel_adsp/ace/Kconfig.defconfig.series [hunk 2]
```diff
@@ -68,6 +60,6 @@ config LOG_BACKEND_ADSP
 
 endif # LOG
 
-source "soc/soc_legacy/xtensa/intel_adsp/ace/Kconfig.defconfig.ace*"
+# rsource "Kconfig.defconfig.ace*"
```
Comment (nordicjm) [inline soc/intel/intel_adsp/ace/Kconfig.defconfig.series:63]: ??

File: soc/intel/intel_adsp/ace/Kconfig.series [hunk 3]
```diff
@@ -14,3 +15,12 @@ config SOC_SERIES_INTEL_ACE
 	select HAS_PM
 	help
 	  Intel ADSP ACE
+
+config SOC_SERIES
+	default "ace" if SOC_SERIES_INTEL_ACE
+
+config SOC_TOOLCHAIN_NAME
+	string
```
Comment (nordicjm) [inline soc/intel/intel_adsp/ace/Kconfig.series:23]:
```suggestion
```

File: soc/intel/intel_adsp/ace/Kconfig.series [hunk 4]
```diff
@@ -1,8 +1,9 @@
-# Copyright (c) 2022 Intel Corporation
+# Copyright (c) 2022-2024 Intel Corporation
```
Comment (nordicjm) [inline soc/intel/intel_adsp/ace/Kconfig.series:1]: This should be Kconfig.soc

File: soc/intel/intel_adsp/ace/Kconfig.series [hunk 5]
```diff
@@ -1,8 +1,9 @@
...
+#
 # SPDX-License-Identifier: Apache-2.0
 
 config SOC_SERIES_INTEL_ACE
-	bool "Intel ACE"
+	bool
 	select SOC_FAMILY_INTEL_ADSP
 	select XTENSA
```
Comment (nordicjm) [inline soc/intel/intel_adsp/ace/Kconfig.series:8]: These lines should be in the Kconfig file

File: soc/intel/intel_adsp/ace/Kconfig.soc [hunk 6]
```diff
@@ -0,0 +1,26 @@
+# Copyright (c) 2022-2024 Intel Corporation
+# SPDX-License-Identifier: Apache-2.0
+
+config SOC_INTEL_ACE15_MTPM
+	bool
+	help
+	  ACE 1.5 Meteor Lake PCH M
```
Comment (nordicjm) [inline soc/intel/intel_adsp/ace/Kconfig.soc:7]: help goes at bottom

File: soc/intel/intel_adsp/ace/Kconfig.soc [hunk 7]
```diff
@@ -0,0 +1,26 @@
...
+	select SOC_SERIES_INTEL_ACE
+
+config SOC_INTEL_ACE20_LNL
+	bool
+	help
```
Comment (nordicjm) [inline soc/intel/intel_adsp/ace/Kconfig.soc:12]: etc.

File: soc/intel/intel_adsp/cavs/Kconfig.series [hunk 8]
```diff
@@ -1,8 +1,9 @@
-# Copyright (c) 2017,2022 Intel Corporation
+# Copyright (c) 2017,2022-2024 Intel Corporation
```
Comment (nordicjm) [inline soc/intel/intel_adsp/cavs/Kconfig.series:1]: All `Kconfig.series` files should be renamed to `Kconfig.soc`

File: boards/intel/intel_adsp_cavs25/Kconfig.intel_adsp_cavs25 [hunk 9]
```diff
@@ -0,0 +1,5 @@
+# Copyright (c) 2024 Intel Corporation
+# SPDX-License-Identifier: Apache-2.0
+
+config  BOARD_INTEL_ADSP_CAVS25
```
Comment (nordicjm) [inline boards/intel/intel_adsp_cavs25/Kconfig.intel_adsp_cavs25:4]: double space after Kconfig

File: boards/intel/intel_adsp_cavs25/Kconfig.intel_adsp_cavs25_tgph [hunk 10]
```diff
@@ -0,0 +1,5 @@
+# Copyright (c) 2024 Intel Corporation
+# SPDX-License-Identifier: Apache-2.0
+
+config  BOARD_INTEL_ADSP_CAVS25_TGPH
```
Comment (nordicjm) [inline boards/intel/intel_adsp_cavs25/Kconfig.intel_adsp_cavs25_tgph]: etc

**Review by nordicjm — CHANGES_REQUESTED** [review]

**Comment by nashif** [comment]
> As said before, if there is only one soc then you can omit the soc, i.e. `intel_adsp_cavs25//tgph`, that should only be omitted by users on the command line though, .yaml files and sample.yml/testcase.yml should use the full name

as said before, this context aware board naming and guessing is not ideal and is going to be very error-prone. The fact is that for reference we will need one unique and authorative name for boards and should not rely on the context, i.e. for tooling and what not. Such long names with lots of redundancy are not user or developer friendly at all. I am sure this can be solved in the definition of the board or soc itself, but we are giving a people a rope to hang themselves with and we will end up with 1m long board names with lots of redunancy that is not as intuitive as it may seem.

**Comment by nashif** [comment]
@tejlmand  @nordicjm  btw, there is something very wrong here, SOF failing to build and picking kconfigs from some completely unrelated _defconfig file (for CONFIG_MULTICORE defined in SOF). Looks like some handling of ARCH is not quite right.

**Comment by golowanow** [comment]
> @tejlmand @nordicjm btw, there is something very wrong here, SOF failing to build and picking kconfigs from some completely unrelated _defconfig file (for CONFIG_MULTICORE defined in SOF). Looks like some handling of ARCH is not quite right.

the proposed fix will be in this PR soon

File: soc/intel/intel_adsp/Kconfig (hunk 1 above)
Comment (golowanow) [reply soc/intel/intel_adsp/Kconfig]: changed

File: soc/intel/intel_adsp/ace/Kconfig.defconfig.series (hunk 2 above)
Comment (golowanow) [reply soc/intel/intel_adsp/ace/Kconfig.defconfig.series]: my bad, left while debugging

File: soc/intel/intel_adsp/ace/Kconfig.series (hunk 5 above)
Comment (golowanow) [reply soc/intel/intel_adsp/ace/Kconfig.series]: changed

File: soc/intel/intel_adsp/ace/Kconfig.soc (hunk 6 above)
Comment (golowanow) [reply soc/intel/intel_adsp/ace/Kconfig.soc:20]: done

File: soc/intel/intel_adsp/ace/Kconfig.soc (hunk 7 above)
Comment (golowanow) [reply soc/intel/intel_adsp/ace/Kconfig.soc:25]: done

File: soc/intel/intel_adsp/cavs/Kconfig.series (hunk 8 above)
Comment (golowanow) [reply soc/intel/intel_adsp/cavs/Kconfig.series]: changed

File: boards/intel/intel_adsp_cavs25/Kconfig.intel_adsp_cavs25 (hunk 9 above)
Comment (golowanow) [reply boards/intel/intel_adsp_cavs25/Kconfig.intel_adsp_cavs25]: done

File: boards/intel/intel_adsp_cavs25/Kconfig.intel_adsp_cavs25_tgph (hunk 10 above)
Comment (golowanow) [reply boards/intel/intel_adsp_cavs25/Kconfig.intel_adsp_cavs25_tgph]: done

File: soc/intel/intel_adsp/Kconfig.defconfig
Comment (golowanow) [inline soc/intel/intel_adsp/Kconfig.defconfig:1]: where CORE_COUNT is set to default 1 at SOF module when its Kconfig is processed by the HWMv2 altered build sequence (without ARCH set):
https://github.com/zephyrproject-rtos/sof/blob/0606152d4aafc1f7ed43df1b1813252bfc74e154/src/arch/host/Kconfig#L5-L9

so the workaround's idea is to just re-align to MAX_CORE_COUNT of the SoC similar to
https://github.com/zephyrproject-rtos/sof/blob/0606152d4aafc1f7ed43df1b1813252bfc74e154/src/platform/Kconfig#L240-L247

**Review by nordicjm — APPROVED** [review]

**Comment by golowanow** [comment]
@tejlmand, @nashif, @dcpleung - up to your attention.

**Comment by golowanow** [comment]
adjusted to the `collab-hwm` rebase

**Comment by golowanow** [comment]
fix `KconfigHWMv2` check issues of this PR, but still needs #68811 to run it on CI

**Comment by nashif** [comment]
Thought about it a bit and played with this PR, I think it makes much more sense to have a revised structure that will give us something like this:

```

  Valid board identifiers for intel_adsp are:

  intel_adsp/cavs25

  intel_adsp/cavs25/tgph

  intel_adsp/ace15_mtpm

  intel_adsp/ace20_lnl
```

So basically have the board definition go under intel_adsp and revise soc name to match v1 naming, this will give us a very close naming to what we had before and will make it future proof.

**Comment by golowanow** [comment]
rebase and few more path changes in cavs25 documentation and west runner script.

**Comment by golowanow** [comment]
> I think it makes much more sense to have a revised structure that will give us something like this:
> 
> ```
>   intel_adsp/cavs25
>   intel_adsp/cavs25/tgph
> ```
> 
> So basically have the board definition go under intel_adsp and revise soc name to match v1 naming, this will give us a very close naming to what we had before and will make it future proof.

@nashif  I've added few commits to rename `cavs25` is it what you think we should do for it,  same for `ace` ?
@nordicjm can you also look ?

**Comment by nashif** [comment]
> and now it can't run on CI here because of platform name changed:
> 
> ```
> ERROR   - sample.audio.sof - integration_platforms - unrecognized platform - intel_adsp_cavs25
> ```

that is expected, we need to change the yaml file in sof, so submit a corrosponding PR to the sof project and link it in here in optional manifest using the PR to https://github.com/zephyrproject-rtos/sof

File: submanifests/optional.yaml [hunk 11]
```diff
@@ -34,7 +34,7 @@ manifest:
       groups:
         - optional
     - name: sof
-      revision: 0606152d4aafc1f7ed43df1b1813252bfc74e154
+      revision: b9d637c1b31e597064b7b946f4f18fb73e3dca5b
```
Comment (nashif) [inline submanifests/optional.yaml:37]: you should be pointing to a PR, not a commit sha

**Comment by golowanow** [comment]
> > I think it makes much more sense to have a revised structure that will give us something like this:
> > ```
> >   intel_adsp/cavs25
> >   intel_adsp/cavs25/tgph
> > ```
> >     
> >   
> > So basically have the board definition go under intel_adsp and revise soc name to match v1 naming, this will give us a very close naming to what we had before and will make it future proof.
> 
> @nashif I've added few commits to rename `cavs25` is it what you think we should do for it, same for `ace` ? @nordicjm can you also look at this ?

@nashif so moving all boards to this scheme or keep v1 for now  ?

**Comment by golowanow** [comment]
> > >
```
  intel_adsp/cavs25
  intel_adsp/cavs25/tgph
  intel_adsp/ace15_mtpm
  intel_adsp/ace20_lnl
``` 
> > > So basically have the board definition go under intel_adsp and revise soc name to match v1 naming, this will give us a very close naming to what we had before and will make it future proof.
> > 
> 
> @nashif so moving all boards to this scheme or keep v1 for now ?

all `intel_adsp` boards are renamed to the proposed shortened form.

**Comment by golowanow** [comment]
re-assign reviewers according to the current MANIFEST

File: boards/intel/intel_adsp/Kconfig.intel_adsp [hunk 12]
```diff
@@ -0,0 +1,18 @@
...
+	bool
+	select SOC_INTEL_CAVS_V25
+
+config BOARD_INTEL_ADSP_ACE15_MTPM
+	bool
+	select SOC_INTEL_ACE15_MTPM
+
+config BOARD_INTEL_ADSP_ACE20_LNL
+	bool
+	select SOC_INTEL_ACE20_LNL
```
Comment (tejlmand) [inline boards/intel/intel_adsp/Kconfig.intel_adsp:18]: Please, no bool, as the `BOARD_<name>` and `BOARD_<full_identifier>` are constructed here:
https://github.com/zephyrproject-rtos/zephyr/blob/db742d03aa8fa469a96435fe85b292e62d061d38/boards/Kconfig.v2#L8
https://github.com/zephyrproject-rtos/zephyr/blob/db742d03aa8fa469a96435fe85b292e62d061d38/boards/Kconfig.v2#L13

Some details can be found here: https://docs.zephyrproject.org/latest/build/kconfig/tips.html#common-kconfig-shorthands

Please do:
```suggestion
config BOARD_INTEL_ADSP
	select SOC_INTEL_CAVS_V25 if BOARD_INTEL_ADSP_CAVS25
	select SOC_INTEL_CAVS_V25 if BOARD_INTEL_ADSP_CAVS25_TGPH
	select SOC_INTEL_ACE15_MTPM if BOARD_INTEL_ADSP_ACE15_MTPM
	select SOC_INTEL_ACE20_LNL if  BOARD_INTEL_ADSP_ACE20_LNL
```
as that will ensure any typos or similar wrong selection is caught by CI.

and it's fewer lines 😉

**Review by tejlmand — CHANGES_REQUESTED** [review]
Thanks for this porting, but please update the board's Kconfig.

File: boards/intel/intel_adsp/Kconfig.intel_adsp (hunk 12 above)
Comment (golowanow) [reply boards/intel/intel_adsp/Kconfig.intel_adsp:18]: > Please, no bool, as the `BOARD_<name>` and `BOARD_<full_identifier>` are constructed here:

well, as I recall, without these `bool` it fails on build as config variables has no type for other boards except the current one.
@tejlmand m.b. there is some other problem with this way of board definition change then ?

Comment (golowanow) [reply boards/intel/intel_adsp/Kconfig.intel_adsp]: > and it's fewer lines 😉

right, that's it, thank you, changed.

File: boards/intel/intel_adsp/board.cmake [hunk 13]
```diff
@@ -0,0 +1,42 @@
+# Copyright (c) 2022-2024 Intel Corporation
+#
+# SPDX-License-Identifier: Apache-2.0
+
+if(CONFIG_BOARD_INTEL_ADSP_CAVS25 OR CONFIG_BOARD_INTEL_ADSP_CAVS25_TGPH)
```
[the remark this answers is not in the document]
Comment (golowanow) [reply boards/intel/intel_adsp/board.cmake:5]: it has been re-composed from 3 `board.cmake` files, so it is rather created anew.

[the remark this answers is not in the document]
Comment (tejlmand) [reply boards/intel/intel_adsp/board.cmake:5]: > it has been re-composed from 3 `board.cmake` files, so it is rather created anew.

in that case fixing indentation would be nice.

One more case of `git` not really supporting renames, but instead relies on the similarity index matching 😉 
```
diff --git a/boards/intel/intel_adsp_cavs25/board.cmake b/boards/intel/intel_adsp/board.cmake
similarity index 90%
rename from boards/intel/intel_adsp_cavs25/board.cmake
rename to boards/intel/intel_adsp/board.cmake
index 46692b0b028..1bdb2698c12 100644
--- a/boards/intel/intel_adsp_cavs25/board.cmake
+++ b/boards/intel/intel_adsp/board.cmake
...
```

**Review by tejlmand — APPROVED** [review]
approved, but as it turns out `board.cmake` is created new, then fixing indentation in the process would be nice, but can also be done later.

**Comment by golowanow** [comment]
> approved, but as it turns out `board.cmake` is created new, then fixing indentation in the process would be nice, but can also be done later.

indentation fixed

**Review by nashif — APPROVED** [review]

File: soc/intel/intel_adsp/Kconfig.defconfig [hunk 14]
```diff
@@ -1,11 +1,18 @@
...
-source "soc/soc_legacy/xtensa/intel_adsp/*/Kconfig.defconfig.series"
 
+rsource "*/Kconfig.defconfig.series"
+
+# A workaround for HWMv2 to recover SOF arch/xtensa defaults overridden by arch/host.
+if SOF
+config CORE_COUNT
+	int
+	default MP_MAX_NUM_CPUS
+endif
```
Comment (ceolin) [inline soc/intel/intel_adsp/Kconfig.defconfig:15]: Can't we get rid of this now since we already have to do changes in SOF side ?

**Pull request merged**